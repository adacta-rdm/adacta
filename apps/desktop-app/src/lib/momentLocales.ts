import moment from "moment";

// Bundle de
import "moment/dist/locale/de.js";
moment.locale("de");
