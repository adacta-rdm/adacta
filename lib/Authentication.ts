import { pbkdf2 as _pbkdf2, randomBytes } from "crypto";
import { promisify } from "util";

const pbkdf2 = promisify(_pbkdf2);

const hashIterations = 100000;

async function hashPassword(password: string, salt: string) {
	const buffer = await pbkdf2(password, salt, hashIterations, 64, "sha512");
	return buffer.toString("base64");
}

export async function verifyPassword(password: string, hash: string, salt: string) {
	const newHash = await hashPassword(password, salt);
	return newHash === hash;
}

export async function createPasswordHash(password: string) {
	const salt = randomBytes(128).toString("base64");
	const passwordHash = await hashPassword(password, salt);

	return { salt, passwordHash };
}
