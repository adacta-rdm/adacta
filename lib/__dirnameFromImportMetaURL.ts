/**
 * Pass in `import.meta.url` as the argument and it will return the directory name of the current, similar to
 * `__dirname` in commonjs.
 *
 * Avoids the use of Node APIs so that it also works in the browser.
 */
export function __dirnameFromImportMetaURL(importMetaURL: string): string {
	const { pathname } = new URL(importMetaURL);
	const lastSlash = pathname.lastIndexOf("/");

	return pathname.slice(0, lastSlash);
}
