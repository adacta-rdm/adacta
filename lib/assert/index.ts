/**
 * @file Automatically generated by barrelsby.
 */

export * from "./assertDefined";
export * from "./assertInstanceof";
export * from "./assertUnreachable";
export * from "./isNonNullish";
