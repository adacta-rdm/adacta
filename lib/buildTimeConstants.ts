/**
 * The constants defined in this file are embedded into the bundles at build time using the __ADACTA global.
 * During development, the __ADACTA global is not defined, so the constants are defined dynamically.
 *
 * This approach allows us to also bundle text files into the build, while maintaining compatibility with tsx which is
 * used in development (while tsx uses esbuild under the hood, it does not support esbuild loaders).
 */
import { readFileSync } from "node:fs";

// NOTE: This import should not be shortened to "~/dev/sh" because it is used in the build process.
import { shSync } from "../dev/sh";
import { version } from "../package.json";

globalThis.__ADACTA = globalThis.__ADACTA ?? {};

export const COMMIT_SHA =
	__ADACTA.COMMIT_SHA ?? shSync("git rev-parse --short HEAD").trim().slice(0, 8);
export const BUILD_DATE = __ADACTA.BUILD_DATE ?? new Date().toISOString().slice(0, 10);

/**
 * The version specified in the package.json file.
 */
export const PACKAGE_VERSION = __ADACTA.PACKAGE_VERSION ?? version;

export const TYPE_DEFS =
	__ADACTA.TYPE_DEFS ??
	readFileSync(`${__dirname}/../apps/repo-server/src/graphql/generated/schema.graphql`, "utf-8");

export const INIT_REPO_SQL =
	__ADACTA.INIT_REPO_SQL ?? readFileSync(`${__dirname}/../drizzle/init-repo.sql`, "utf-8");

export const INIT_MIGRATIONS_SQL =
	__ADACTA.INIT_MIGRATIONS_SQL ??
	readFileSync(`${__dirname}/../drizzle/init-migrations.sql`, "utf-8");

declare global {
	// The value of the Record type is typed as undefined, so the actually exported value gets the type of the
	// default value (i.e. the part after the ?? operator).
	// This needs to be declared using var...
	// eslint-disable-next-line no-var
	var __ADACTA: Record<string, undefined>;
}
