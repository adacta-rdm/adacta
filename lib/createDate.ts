import type { Opaque } from "type-fest";

const ISO_REGEX =
	/(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/;

export type IDatetime = Opaque<string, Date>;

export function assertIsoDate(isoString: string): asserts isoString is IDatetime {
	if (!ISO_REGEX.test(isoString)) {
		throw new Error(`'${isoString}' is not an ISO Date string`);
	}
}

export function createMaybeDate(maybeIsoString: string | null | undefined) {
	if (maybeIsoString === null || maybeIsoString === undefined) {
		return undefined;
	}
	return createDate(maybeIsoString);
}

export function createDate(isoString: string) {
	assertIsoDate(isoString);
	return new Date(isoString);
}

export function createMaybeIDatetime(
	date: Date | IDatetime | null | undefined
): IDatetime | undefined {
	if (date === null || date === undefined) {
		return undefined;
	}
	return createIDatetime(date);
}

export function createIDatetime(date: Date | IDatetime): IDatetime {
	if (date instanceof Date) return date.toISOString() as IDatetime;
	return date;
}
