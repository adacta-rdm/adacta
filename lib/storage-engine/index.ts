/**
 * @file Automatically generated by barrelsby.
 */

export * from "./FileNotFoundError";
export * from "./FileSystemStorageEngine";
export * from "./ReusableFileHandle";
export * from "./S3StorageEngine";
export * from "./StorageEngine";
export * from "./refreshableTimeout";
