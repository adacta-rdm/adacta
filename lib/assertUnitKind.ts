import { AssertionError } from "assert";

import { Qty } from "@omegadot/einheiten";
import type { UnitKind } from "@omegadot/einheiten/dist/types/quantities/kind";
import { difference } from "lodash";

export function assertUnitKind(kind: string): asserts kind is UnitKind {
	if (!Qty.getKinds().includes(kind as UnitKind)) {
		throw new AssertionError({ message: `Unit kind ${kind} is not a valid unit kind` });
	}
}

export function assertUnitKinds(kind: string[] | Readonly<string[]>): asserts kind is UnitKind[] {
	const inputUnique = [...new Set(kind)];
	const validKinds = Qty.getKinds();

	const invalidUnits = difference(inputUnique, validKinds);

	if (invalidUnits.length > 0) {
		throw new AssertionError({
			message: `The following unit kinds are not valid unit kinds: ${invalidUnits.join(",")}`,
		});
	}
}
