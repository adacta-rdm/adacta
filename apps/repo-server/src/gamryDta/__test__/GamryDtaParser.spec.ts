import { describe, expect, test } from "vitest";

import type {
	GamryDtaParserEvents,
	TGamryMetadata,
} from "~/apps/repo-server/src/gamryDta/GamryFileReader";
import {
	GamryFileReader,
	OverflowParser,
	OverflowTypes,
} from "~/apps/repo-server/src/gamryDta/GamryFileReader";
import { FileSystemStorageEngine } from "~/lib/storage-engine";
import type { Readable } from "~/lib/streams";

describe("GamryDtaParser", () => {
	test.each([
		{ filename: "ocp_data.dta", lengths: [11] },
		{ filename: "eispot_data.dta", lengths: [3] },
		{ filename: "cv_data.dta", lengths: [3, 3, 3, 3] },
	])("$filename", async (testObj) => {
		const { filename, lengths } = testObj;

		const sto = new FileSystemStorageEngine(`${__dirname}/data`);
		const gamryParser = new GamryFileReader(sto, filename);

		const datastream = await gamryParser.parse();
		const result = await consumeParserStreamToObject(datastream);

		expect(result.tables).toHaveLength(lengths.length);
		expect(result.tables.map((t) => t.length)).toStrictEqual(lengths);
		expect(result.expectedSizes).toMatchSnapshot(); // Some tables have expected sizes, some don't
	});

	describe("Partial file", () => {
		test("cv_data_partial.dta", async () => {
			const sto = new FileSystemStorageEngine(`${__dirname}/data`);
			const gamryParser = new GamryFileReader(sto, "cv_data_partial.dta");

			const datastream = await gamryParser.parse();
			await expect(datastream.promise()).rejects.toThrow();
		});
	});

	describe("OverflowParser", () => {
		test("isOverflowFlagSet", () => {
			expect(OverflowParser.isOverflowFlagSet("..........a", OverflowTypes.ADC_AUX)).toBe(true);
			expect(OverflowParser.isOverflowFlagSet("..........a", OverflowTypes.ADC_RANGE_I)).toBe(
				false
			);

			expect(OverflowParser.isOverflowFlagSet("......s...a", OverflowTypes.ADC_AUX)).toBe(true);
			expect(
				OverflowParser.isOverflowFlagSet("......s...a", OverflowTypes.SETTELING_PROBLEM_HARDWARE)
			).toBe(true);
		});
	});
});

function consumeParserStreamToObject(datastream: Readable<GamryDtaParserEvents>) {
	let tableCounter = -1;

	const tables: (string | number)[][][] = []; // dimension 1: table, dimension 2: row, dimension 3: column
	const minMax: { min: (number | undefined)[]; max: (number | undefined)[] }[] = [];

	let metadata!: TGamryMetadata;
	let auxData!: string[];
	let displayNames!: Map<string, string>;

	const expectedSizes: (number | undefined)[] = [];

	return new Promise<{
		metadata: TGamryMetadata | undefined;
		auxData: string[];
		displayNames: Map<string, string>;
		tables: typeof tables;
		expectedSizes: typeof expectedSizes;
		minMax: typeof minMax;
	}>((resolve) => {
		datastream.on("data", (data) => {
			if (data.type === "metadata") {
				metadata = data.data.metadata;
				auxData = data.data.auxData;
				displayNames = data.data.displayNames;
			}

			if (data.type === "next") {
				tableCounter++;
				expectedSizes[tableCounter] = data.expectedCurveLength;
			}

			if (data.type === "data") {
				tables[tableCounter] = tables.at(tableCounter) ?? [];
				tables[tableCounter].push(data.data);
			}

			if (data.type === "minmax") {
				minMax[tableCounter] = { min: data.min, max: data.max };
			}
		});

		datastream.on("end", () => {
			resolve({ metadata, auxData, displayNames, tables, expectedSizes, minMax });
		});
	});
}
