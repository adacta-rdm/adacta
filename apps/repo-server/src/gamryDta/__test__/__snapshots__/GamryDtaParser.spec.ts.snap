// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GamryDtaParser > 'cv_data.dta' 1`] = `
[
  undefined,
  undefined,
  undefined,
  undefined,
]
`;

exports[`GamryDtaParser > 'eispot_data.dta' 1`] = `
[
  undefined,
]
`;

exports[`GamryDtaParser > 'ocp_data.dta' 1`] = `
[
  11,
]
`;

exports[`GamryDtaParser > RDE 'cv_data.dta' 1`] = `
[
  undefined,
  undefined,
  undefined,
  undefined,
]
`;

exports[`GamryDtaParser > RDE 'cv_real_metadata.dta' 1`] = `
[
  undefined,
  undefined,
  undefined,
  undefined,
]
`;

exports[`GamryDtaParser > RDE 'eispot_data.dta' 1`] = `
[
  undefined,
]
`;

exports[`GamryDtaParser > RDE 'ocp_data.dta' 1`] = `
[
  11,
]
`;

exports[`GamryDtaParser > chronoa_data.dta 1`] = `
{
  "auxData": [],
  "displayNames": Map {
    "tag" => "CHRONOA",
    "title" => "Test &Identifier",
    "date" => "Date",
    "time" => "Time",
    "notes" => "&Notes...",
    "tprestep" => "Pre-step Delay &Time (s)",
    "vstep1" => "Step 1 Voltage (V)",
    "tstep1" => "Step 1 Time (s)",
    "vstep2" => "Step 2 Voltage (V)",
    "tstep2" => "Step 2 Time (s)",
    "sampletime" => "Sa&mple Period (s)",
    "ilimit" => "Limit &I (mA/cm^2)",
    "imode" => "I/E Range &Mode",
    "irange" => "&Max Current (mA)",
    "decimate" => "Decimate",
    "area" => "Electrode &Area (cm^2)",
    "electrodetype" => "Electrode Type",
    "stripping" => "Used for Stripping",
    "polarity" => "Signal Polarity",
    "eqdelay" => "Equil. &Time (s)",
    "condit" => "E(V)",
    "delay" => "Stab.(mV/s)",
    "ircomp" => "IRCom&p",
    "pfcor" => "PF Corr. (ohm)",
    "samplingmode" => "Sampling &Mode",
    "eoc" => "Open Circuit (V)",
    "pstatmodel" => "Pstat Model",
    "pstatsection" => "Pstat Section",
    "pstatserialno" => "Pstat Serial Number",
    "ctrlmode" => "Control Mode",
    "iestab" => "I/E Stability",
    "caspeed" => "Control Amp Speed",
    "convention" => "Current Convention",
    "ichrange" => "Ich Range",
    "ichrangemode" => "Ich Auto Range",
    "ichoffsetenable" => "Ich Offset Enable",
    "ichoffset" => "Ich Offset (V)",
    "ichfilter" => "Ich Filter",
    "vchrange" => "Vch Range",
    "vchrangemode" => "Vch Auto Range",
    "vchoffsetenable" => "Vch Offset Enable",
    "vchoffset" => "Vch Offset (V)",
    "vchfilter" => "Vch Filter",
    "achrange" => "Ach Range",
    "achoffsetenable" => "Ach Offset Enable",
    "achoffset" => "Ach Offset (V)",
    "achfilter" => "Ach Filter",
    "ierangelowerlimit" => "I/E Range Lower Limit",
    "ierangemode" => "I/E AutoRange",
    "ierange" => "I/E Range",
    "posfeedenable" => "Positive Feedback IR Comp",
    "posfeedresistance" => "Positive Feedback Resistance (ohm)",
    "achselect" => "Ach Select",
    "cableid" => "Cable ID",
    "dccaldate" => "DC Calibration Date",
    "accaldate" => "AC Calibration Date",
    "thermoselect" => "Thermo Select",
    "frameworkversion" => "Framework Version",
    "instrumentversion" => "Instrument Version",
  },
  "expectedSizes": [
    5258,
  ],
  "metadata": {
    "accaldate": "3/10/2019",
    "achfilter": 2,
    "achoffset": 0,
    "achoffsetenable": false,
    "achrange": 2,
    "achselect": 0,
    "area": 1,
    "cableid": 14,
    "caspeed": 2,
    "condit": {
      "enable": false,
      "end": 0,
      "start": 15,
    },
    "convention": 1,
    "ctrlmode": 1,
    "date": "3/10/2019",
    "dccaldate": "3/10/2019",
    "decimate": false,
    "delay": {
      "enable": false,
      "end": 0.1,
      "start": 300,
    },
    "electrodetype": 0,
    "eoc": 0,
    "eqdelay": 5,
    "frameworkversion": 0,
    "ichfilter": 3,
    "ichoffset": 0,
    "ichoffsetenable": false,
    "ichrange": 2,
    "ichrangemode": true,
    "ierange": 6,
    "ierangelowerlimit": 1,
    "ierangemode": false,
    "iestab": 0,
    "ilimit": 1,
    "imode": 1,
    "instrumentversion": "0.00",
    "irange": 0.001,
    "ircomp": 0,
    "notes": "PSTAT	PSTAT	potentiostat-id	Potentiostat",
    "pfcor": 50,
    "polarity": false,
    "posfeedenable": false,
    "posfeedresistance": 0,
    "pstatmodel": 4,
    "pstatsection": "potentiostat-id",
    "pstatserialno": "00000",
    "sampletime": 30,
    "samplingmode": 1,
    "stripping": false,
    "tag": "CHRONOA",
    "thermoselect": 1,
    "time": "12:00:00",
    "title": "Experiment-Title",
    "tprestep": 0.5,
    "tstep1": 30,
    "tstep2": 1000000,
    "vchfilter": 3,
    "vchoffset": 0,
    "vchoffsetenable": false,
    "vchrange": 3,
    "vchrangemode": true,
    "vstep1": 0.5,
    "vstep2": 0.5,
  },
  "minMax": [
    {
      "max": [],
      "min": [],
    },
  ],
  "tables": [
    [
      [
        0,
        0,
        -0.00054,
        -2.34197e-8,
        0,
        0,
        -0.000666902,
        6,
        "...........",
      ],
      [
        1,
        30,
        0.499669,
        1.40441e-8,
        0,
        0.5,
        -0.000664236,
        6,
        "...........",
      ],
      [
        2,
        60,
        0.499668,
        9.32775e-9,
        0,
        0.5,
        -0.000663354,
        6,
        "...........",
      ],
      [
        3,
        90.0001,
        0.499659,
        7.52058e-9,
        0,
        0.5,
        -0.000665815,
        6,
        "...........",
      ],
      [
        4,
        120,
        0.499654,
        6.37081e-9,
        0,
        0.5,
        -0.000665403,
        6,
        "...........",
      ],
      [
        5,
        150,
        0.499657,
        5.4468e-9,
        0,
        0.5,
        -0.000665363,
        6,
        "...........",
      ],
      [
        6,
        180,
        0.499661,
        4.86946e-9,
        0,
        0.5,
        -0.00066566,
        6,
        "...........",
      ],
      [
        7,
        210,
        0.499667,
        4.43861e-9,
        0,
        0.5,
        -0.000664322,
        6,
        "...........",
      ],
      [
        8,
        240,
        0.49966,
        4.07718e-9,
        0,
        0.5,
        -0.000665789,
        6,
        "...........",
      ],
      [
        9,
        270,
        0.4,
        3e-9,
        0,
        0.5,
        -0.000667797,
        6,
        "...........",
      ],
    ],
  ],
}
`;

exports[`GamryDtaParser > cv_data.dta 1`] = `
{
  "auxData": [],
  "displayNames": Map {
    "tag" => "CV",
    "title" => "Test &Identifier",
    "date" => "Date",
    "time" => "Time",
    "checknotes" => "&Notes...",
    "checkpoten" => "Initial &E (V)",
    "checkquant" => "&Scan Rate (mV/s)",
    "checkiquant" => "C&ycles (#)",
    "checkselector" => "I/E Range &Mode",
    "checktoggle" => "Used for Stripping",
    "check2param" => "E(V)",
    "vlimit1" => "Scan Limit &1 (V)",
    "vlimit2" => "Scan Limit &2 (V)",
    "scanrate" => "&Scan Rate (mV/s)",
    "condit" => "E(V)",
    "delay" => "Stab.(mV/s)",
  },
  "expectedSizes": [
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  "metadata": {
    "check2param": {
      "enable": true,
      "end": 0.5,
      "start": 300,
    },
    "checkiquant": 5,
    "checknotes": "test-notes-data",
    "checkpoten": 0.5,
    "checkquant": 1.2345,
    "checkselector": 0,
    "checktoggle": false,
    "condit": {
      "enable": false,
      "end": 0.4,
      "start": 300,
    },
    "date": "3/6/2019",
    "delay": {
      "enable": false,
      "end": 0.1,
      "start": 300,
    },
    "scanrate": 1.23456,
    "tag": "CV",
    "time": "16:35:22",
    "title": "Cyclic Voltammetry",
    "vlimit1": 0.1,
    "vlimit2": 0.9,
  },
  "minMax": [
    {
      "max": [
        9,
        1,
        0.5,
        7.80498e-9,
        0,
        0.5,
        -0.00106785,
        5,
      ],
      "min": [
        0,
        0.1,
        0.49,
        -2.41658e-9,
        0,
        0.491,
        -0.00109885,
        5,
      ],
    },
    {
      "max": [
        19,
        121.1,
        0.897987,
        6.57772e-7,
        0,
        0.899,
        -0.00107685,
        6,
      ],
      "min": [
        10,
        120.2,
        0.889031,
        5.25977e-7,
        0,
        0.89,
        -0.00109995,
        6,
      ],
    },
    {
      "max": [
        29,
        281.1,
        0.897846,
        4.74862e-7,
        0,
        0.899,
        -0.00107575,
        5,
      ],
      "min": [
        20,
        280.2,
        0.888906,
        3.69861e-7,
        0,
        0.89,
        -0.00110415,
        5,
      ],
    },
    {
      "max": [
        39,
        441.1,
        0.897891,
        3.98725e-7,
        0,
        0.899,
        -0.00106645,
        5,
      ],
      "min": [
        30,
        440.2,
        0.88896,
        3.02352e-7,
        0,
        0.89,
        -0.00110165,
        5,
      ],
    },
    {
      "max": [
        49,
        601.1,
        0.897929,
        3.52667e-7,
        0,
        0.899,
        -0.00107535,
        5,
      ],
      "min": [
        40,
        600.2,
        0.889001,
        2.62272e-7,
        0,
        0.89,
        -0.00110035,
        5,
      ],
    },
  ],
  "tables": [
    [
      [
        0,
        0.1,
        0.49,
        7.80498e-9,
        0,
        0.5,
        -0.00107185,
        5,
        "...........",
      ],
      [
        1,
        0.2,
        0.498267,
        5.33521e-9,
        0,
        0.499,
        -0.00107255,
        5,
        "...........",
      ],
      [
        2,
        0.3,
        0.497266,
        3.79513e-9,
        0,
        0.498,
        -0.00108405,
        5,
        "...........",
      ],
      [
        3,
        0.4,
        0.496276,
        2.55585e-9,
        0,
        0.497,
        -0.00107205,
        5,
        "...........",
      ],
      [
        4,
        0.5,
        0.495274,
        1.43084e-9,
        0,
        0.496,
        -0.00108195,
        5,
        "...........",
      ],
      [
        5,
        0.6,
        0.494277,
        5.3139e-10,
        0,
        0.495,
        -0.00108425,
        5,
        "...........",
      ],
      [
        6,
        0.7,
        0.493266,
        -3.10842e-10,
        0,
        0.494,
        -0.00108675,
        5,
        "...........",
      ],
      [
        7,
        0.8,
        0.492278,
        -1.02199e-9,
        0,
        0.493,
        -0.00107365,
        5,
        "...........",
      ],
      [
        8,
        0.9,
        0.491287,
        -1.75494e-9,
        0,
        0.492,
        -0.00109885,
        5,
        "...........",
      ],
      [
        9,
        1,
        0.5,
        -2.41658e-9,
        0,
        0.491,
        -0.00106785,
        5,
        "...........",
      ],
    ],
    [
      [
        10,
        120.2,
        0.897987,
        6.57772e-7,
        0,
        0.899,
        -0.00108785,
        6,
        "...........",
      ],
      [
        11,
        120.3,
        0.897015,
        6.40211e-7,
        0,
        0.898,
        -0.00108115,
        6,
        "...........",
      ],
      [
        12,
        120.4,
        0.896017,
        6.23702e-7,
        0,
        0.897,
        -0.00109995,
        6,
        "...........",
      ],
      [
        13,
        120.5,
        0.895012,
        6.07876e-7,
        0,
        0.896,
        -0.00108035,
        6,
        "...........",
      ],
      [
        14,
        120.6,
        0.893996,
        5.92887e-7,
        0,
        0.895,
        -0.00109225,
        6,
        "...........",
      ],
      [
        15,
        120.7,
        0.893013,
        5.78426e-7,
        0,
        0.894,
        -0.00108555,
        6,
        "...........",
      ],
      [
        16,
        120.8,
        0.892005,
        5.64491e-7,
        0,
        0.893,
        -0.00108395,
        6,
        "...........",
      ],
      [
        17,
        120.9,
        0.891041,
        5.51307e-7,
        0,
        0.892,
        -0.00107685,
        6,
        "...........",
      ],
      [
        18,
        121,
        0.890015,
        5.38367e-7,
        0,
        0.891,
        -0.00108545,
        6,
        "...........",
      ],
      [
        19,
        121.1,
        0.889031,
        5.25977e-7,
        0,
        0.89,
        -0.00107845,
        6,
        "...........",
      ],
    ],
    [
      [
        20,
        280.2,
        0.897846,
        4.74862e-7,
        0,
        0.899,
        -0.00108665,
        5,
        "...........",
      ],
      [
        21,
        280.3,
        0.896833,
        4.6007e-7,
        0,
        0.898,
        -0.00108555,
        5,
        "...........",
      ],
      [
        22,
        280.4,
        0.895861,
        4.46735e-7,
        0,
        0.897,
        -0.00108645,
        5,
        "...........",
      ],
      [
        23,
        280.5,
        0.894852,
        4.34115e-7,
        0,
        0.896,
        -0.00109015,
        5,
        "...........",
      ],
      [
        24,
        280.6,
        0.89388,
        4.22256e-7,
        0,
        0.895,
        -0.00108125,
        5,
        "...........",
      ],
      [
        25,
        280.7,
        0.892892,
        4.10826e-7,
        0,
        0.894,
        -0.00110415,
        5,
        "...........",
      ],
      [
        26,
        280.8,
        0.891898,
        4.00024e-7,
        0,
        0.893,
        -0.00107575,
        5,
        "...........",
      ],
      [
        27,
        280.9,
        0.890896,
        3.89569e-7,
        0,
        0.892,
        -0.00109685,
        5,
        "...........",
      ],
      [
        28,
        281,
        0.889907,
        3.79479e-7,
        0,
        0.891,
        -0.00107875,
        5,
        "...........",
      ],
      [
        29,
        281.1,
        0.888906,
        3.69861e-7,
        0,
        0.89,
        -0.00108505,
        5,
        "...........",
      ],
    ],
    [
      [
        30,
        440.2,
        0.897891,
        3.98725e-7,
        0,
        0.899,
        -0.00107875,
        5,
        "...........",
      ],
      [
        31,
        440.3,
        0.896907,
        3.8438e-7,
        0,
        0.898,
        -0.00108125,
        5,
        "...........",
      ],
      [
        32,
        440.4,
        0.895912,
        3.72133e-7,
        0,
        0.897,
        -0.00108165,
        5,
        "...........",
      ],
      [
        33,
        440.5,
        0.89492,
        3.60922e-7,
        0,
        0.896,
        -0.00109025,
        5,
        "...........",
      ],
      [
        34,
        440.6,
        0.893923,
        3.50388e-7,
        0,
        0.895,
        -0.00107475,
        5,
        "...........",
      ],
      [
        35,
        440.7,
        0.89294,
        3.39752e-7,
        0,
        0.894,
        -0.00110165,
        5,
        "...........",
      ],
      [
        36,
        440.8,
        0.891939,
        3.29956e-7,
        0,
        0.893,
        -0.00106645,
        5,
        "...........",
      ],
      [
        37,
        440.9,
        0.890949,
        3.20646e-7,
        0,
        0.892,
        -0.00108885,
        5,
        "...........",
      ],
      [
        38,
        441,
        0.889947,
        3.116e-7,
        0,
        0.891,
        -0.00108115,
        5,
        "...........",
      ],
      [
        39,
        441.1,
        0.88896,
        3.02352e-7,
        0,
        0.89,
        -0.00108655,
        5,
        "...........",
      ],
    ],
    [
      [
        40,
        600.2,
        0.897929,
        3.52667e-7,
        0,
        0.899,
        -0.00108155,
        5,
        "...........",
      ],
      [
        41,
        600.3,
        0.896945,
        3.38852e-7,
        0,
        0.898,
        -0.00108615,
        5,
        "...........",
      ],
      [
        42,
        600.4,
        0.895958,
        3.26884e-7,
        0,
        0.897,
        -0.00108925,
        5,
        "...........",
      ],
      [
        43,
        600.5,
        0.894963,
        3.15955e-7,
        0,
        0.896,
        -0.00108635,
        5,
        "...........",
      ],
      [
        44,
        600.6,
        0.89395,
        3.05753e-7,
        0,
        0.895,
        -0.00107885,
        5,
        "...........",
      ],
      [
        45,
        600.7,
        0.892974,
        2.96182e-7,
        0,
        0.894,
        -0.00110035,
        5,
        "...........",
      ],
      [
        46,
        600.8,
        0.891962,
        2.87036e-7,
        0,
        0.893,
        -0.00107535,
        5,
        "...........",
      ],
      [
        47,
        600.9,
        0.890992,
        2.78501e-7,
        0,
        0.892,
        -0.00109475,
        5,
        "...........",
      ],
      [
        48,
        601,
        0.889976,
        2.70272e-7,
        0,
        0.891,
        -0.00107895,
        5,
        "...........",
      ],
      [
        49,
        601.1,
        0.889001,
        2.62272e-7,
        0,
        0.89,
        -0.00108005,
        5,
        "...........",
      ],
    ],
  ],
}
`;
